File: solutions/code-carl/ch10_monotone_stack/T3_LC0503_NextGreaterElements.java
/**
 * Practice template for:
 * - LeetCode 503. Next Greater Element II
 *   https://leetcode.cn/problems/next-greater-element-ii/
 */
import java.util.ArrayDeque;
import java.util.Arrays;
import java.util.Deque;
import java.util.HashMap;

/**
 * @author Cheng Yihao
 * @version 1.0
 * @date 2024/11/28 20:44
 * @comment
 */
public class T3_LC0503_NextGreaterElements {
    public int[] nextGreaterElements(int[] nums) {
        Deque<Integer> deque = new ArrayDeque<>();
        int len = nums.length;
        int[] res = new int[len];
        Arrays.fill(res, -1);

        for (int i = 0; i < len * 2; i++) {
            while (!deque.isEmpty() && nums[i % len] > nums[deque.peek()]) {
                Integer idx = deque.pop();
                res[idx] = nums[i % len];

            }
            deque.push(i % len);
        }

        return res;
    }
}

File: solutions/python/code-carl/ch10_monotone_stack/t3_lc0503_next_greater_elements.py
"""LeetCode 503. Next Greater Element II
https://leetcode.cn/problems/next-greater-element-ii/
"""

from typing import List


class Solution:
    def nextGreaterElements(self, nums: List[int]) -> List[int]:
        ans = [-1] * len(nums)
        stack: List[int] = []

        for i in range(0, 2 * len(nums)):
            while stack and nums[i % len(nums)] > nums[stack[-1]]:
                ans[stack.pop()] = nums[i % len(nums)]
            stack.append(i % len(nums))

        return ans


if __name__ == "__main__":
    solution = Solution()

    assert solution.nextGreaterElements([1, 2, 1]) == [2, -1, 2]

    assert solution.nextGreaterElements([1, 2, 1]) == [2, -1, 2]

    print("All tests passed.")
